builtin/consul/config_sourcer.go

```diff
@@ -0,0 +1,412 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint-plugin-sdk/docs"
+	pb "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+	"github.com/mitchellh/mapstructure"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// retryInterval is the base retry value
+	retryInterval = 5 * time.Second
+
+	// maximum back off time, this is to prevent
+	// exponential runaway
+	maxBackoffTime = 180 * time.Second
+)
+
+type reqConfig struct {
+	Key        string `hcl:"key,attr"`             // kv path to retrieve
+	Namespace  string `hcl:"namespace,optional"`   // namespace the kv data resides within
+	Partition  string `hcl:"partition,optional"`   // partition the kv data resides within
+	Datacenter string `hcl:"datacenter,optional"`  // datacenter the kv data resides within
+	AllowStale bool   `hcl:"allow_stale,optional"` // whether to perform stale queries against non-leader servers
+}
+
+func (r *reqConfig) cacheKey() string {
+	return fmt.Sprintf("%s/%s/%s/%s", r.Datacenter, r.Partition, r.Namespace, r.Key)
+}
+
+type cachedVal struct {
+	kvPair *api.KVPair
+	cancel context.CancelFunc
+	err    error
+}
+
+// ConfigSourcerConfig is used to configure where to talk to Consul, and from
+// where the KV data is to be retrieved
+type ConfigSourcerConfig struct {
```

Comment:
They're not required to be set as configurations here, because the Consul API will attempt to retrieve the configuration from environment variables as well (not unlike Nomad), and will set reasonable defaults if they're not present. 

https://github.com/hashicorp/consul/blob/main/api/api.go#L439-L441